src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp

```diff
@@ -0,0 +1,731 @@
+{
+  if (!Parent())
+    return true;
+
+  const bool left = this == Parent()->Left();
+
+  if (left)
+    return !Parent()->Hyperplane().Right(other.Bound());
+  else
+    return !Parent()->Hyperplane().Left(other.Bound());
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+template<typename VecType>
+inline bool SpillTree<MetricType, StatisticType, MatType, HyperplaneType,
+    SplitType>::HalfSpaceContains(
+    const VecType& point,
+    typename boost::enable_if<IsVector<VecType> >::type*) const
+{
+  if (!Parent())
+    return true;
+
+  const bool left = this == Parent()->Left();
+  const bool toTheLeft = Parent()->Hyperplane().Left(point);
+
+  return left == toTheLeft;
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+void SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+    SplitNode(std::vector<size_t>& points,
+              const size_t maxLeafSize,
+              const double tau,
+              const double rho)
+{
+  // We need to expand the bounds of this node properly.
+  for (size_t i = 0; i < points.size(); i++)
+    bound |= dataset->cols(points[i], points[i]);
+
+  // Calculate the furthest descendant distance.
+  furthestDescendantDistance = 0.5 * bound.Diameter();
+
+  // Now, check if we need to split at all.
+  if (points.size() <= maxLeafSize)
+  {
+    pointsIndex = new std::vector<size_t>();
+    pointsIndex->swap(points);
+    count = pointsIndex->size();
+    return; // We can't split this.
+  }
+
+  const bool split = SplitType<MetricType, MatType>::SplitSpace(bound,
+      *dataset, points, hyperplane);
+  // The node may not be always split. For instance, if all the points are the
+  // same, we can't split them.
+  if (!split)
+  {
+    pointsIndex = new std::vector<size_t>();
+    pointsIndex->swap(points);
+    count = pointsIndex->size();
+    return; // We can't split this.
+  }
+
+  std::vector<size_t> leftPoints, rightPoints;
+  // Split the node.
+  overlappingNode = SplitPoints(tau, rho, points, leftPoints, rightPoints);
+
+  // We don't need the information in points, so lets clean it.
+  std::vector<size_t>().swap(points);
+
+  // Now we will recursively split the children by calling their constructors
+  // (which perform this splitting process).
+  left = new SpillTree(this, leftPoints, tau, maxLeafSize, rho);
+  right = new SpillTree(this, rightPoints, tau, maxLeafSize, rho);
+
+  // Update count number, to represent the number of descendant points.
+  count = left->NumDescendants() + right->NumDescendants();
```

Comment:
I have a concern about this.  This means that `NumDescendants()` of the root can actually be greater than `dataset.n_cols`, but I think this can cause problems.  If we refactor so that `NumDescendants()` of the root returns `dataset.n_cols`, then `Descendant()` will be a little more complex: you'll just have to rearrange the point indices so that the first `n` elements are the non-shared elements, and possibly hold another member to indicate how many non-shared elements the node has.